rfc/0009-qsl-structure.md

```diff
@@ -0,0 +1,129 @@
+		- Drawbacks
+			-  Everybody just uses the fat jar
+			-  API visibility is low since modules don't usually inter-connect
+				-  Compare Forge's Item.setModelRenderer (devoldified) to using a registry for everything
+			-  The massive amount of modules makes navigating FAPI confusing
+			-  Even though there are small modules, modules are often very tiny--Extensions of Items might be split across Content Registries, Item API, Tool Item API, 
+- Forge
+	-
+	- Benefits
+		- Monolithic API makes adding new features straightforward
+	- Drawbacks
+		- It is more difficult to remove parts of code as Mojang breaks them
+		- Lack of compartmentalization leads to seemingly unrelated APIs being intertwined
+		- No api/impl split makes it difficult to know what part of the code you are actually supposed to interact with
+### Goals
+- Be mobile and quick to update like Fabric
+- Have high API visibility like Forge
+- Find a balance between Forge's confusingly monolithic and Fabric's confusingly modular structures
+- Have teams in charge of different parts of QSL's code
+- Avoid making users download and run code they don't need
+## Explanation
+This is an overview of the top-level structure of QSL. Libraries are included, with the names of the FAPI modules that would be under its scope. The actual structure of the modules in each library is up to the team, and it is not a goal to define how they must structure their modules.
+### Libraries
+Each library would be its own folder in the QSL monorepo, with a dedicated team responsible for triage, review, and accepting/denying pull requests. In contrast to Fabric or Forge's system of only one person signing off for the entire standard library, no authority higher than the people on the team would be required to sign off on the PR before it is merged. More clarification on team structure is expected to come in a future RFC.
+
+Libraries only exist for internal structure--there is no fat jar for, say, all `worldgen` modules, neither is there a fat jar for all of QSL. This also means that QSL and its libraries have no versioning.
+### Modules
+Modules follow the following general rules:
+- Maven group of `org.quiltmc.qsl.$library`
+- Artifact id of `mod-id`
+- Semantic Versioning. Hashes are *not* used to identify modules. 
+- Can only depend on modules from `core` or within their own library.
+
+To modders, most mods would not explicitly depend on libraries or their modules, but Loom would automatically determine the modules to depend on by scanning at compile time, and those modules will be downloaded at runtime by Loader.
+### Top-level Structure
+Keep in mind that the modules listed here are the names of the current Fabric API modules, and the final internal structure of QSL **will not** be the same as Fabric API.
+- `core` - All modules here are implicitly depended on by mods by default, even if they do not use any code from them. Mods can choose to opt-out.
```

Comment:
> Mods can choose to opt-out.

This seems unworkable to me, from a community standpoint. From the Quilt community rules (5, specifically)

> Projects must not be malicious - which includes (but is not limited to) malware, mods that include backdoors, mods that modify Quilt in an attempt to hide themselves from servers they connect to, or mods that harass specific people or groups of people (for example, by automatically kicking them from servers with the mod installed)

While this RFC doesn't mess with the wording of the rule, it does mess with the spirit - we've been working under the assumption that whatever modlist/modid handshake logic would always be present - being able to opt out by simply saying "No, thanks" is antithetical to the goal of giving servers control over the mods that their users are making use of.

Whether this means it needs to be factored out into a separate module and have that be required, or whether there's some other solution - that's fine. But providing machinery directly within Quilt to entirely bypass this technical measure will make it much harder for the community team to do their jobs. While the best approach from that aspect may simply be "also disallow mods that opt-out", it still raises the burden of moderation because it's something we have to actually find a way to check directly.

I'm sure there's a more workable approach than this.

Reply:
I'm going to say the concerns about rule 5 are probably not an issue here regarding the technical process. The ability to opt out of using qsl entirely does not appear to me as circumvention or malicious and is rather implementation detail of a mod.

A mod does not need to opt-in to being included in any metadata handshake, that operates off loader's api.

Reply:
Overall I agree with Glitch's approach - there are a lot of concerns to think of here that aren't necessarily related to technology; Glitch's approach seems like the best middle ground to me.

Modularity is a good thing, yes - I've been preaching it myself actually. The idea of something being used to gate users not being included by default is a tricky one, IMO, since most users only use optimisation mods that are unlikely to require QSL, meaning their experience would be broken by servers that require the packet - so including a checkbox or something in the installer that allows them to opt in when no mods use QSL seems like a good solution.

Of course, that could have its own technical problems - like putting it on the classpath and breaking mods that require other verisons. It's a tricky one

Reply:
> Of course, that could have its own technical problems - like putting it on the classpath and breaking mods that require other verisons. It's a tricky one

I don't see this as much of a concern--we can tell Loader to download it the same way it would if some mod depended on it

Reply:
Since we've found a workable solution that's out of scope of QSL, I'm going to consider this issue solved for now.